### ISSUE TITLE
Error when saving the same doc multiple times in parallel

### ISSUE THREAD

**Issue by GilbertGobbels** [Automattic/mongoose#6456]
**Do you want to request a *feature* or report a *bug*?**
A bug.

**What is the current behavior?**
When asynchrously saving the same document twice, if the document had leftover atomics from before the first save, the second save will cause an error that crashes the node process completely; no matter how much error catching is done. The full error + stack is;
```
TypeError: doc.toObject is not a function
    at node_modules\mongoose\lib\types\documentarray.js:224:27
    at Array.map (<anonymous>)
    at Array.toObject (node_modules\mongoose\lib\types\documentarray.js:222:17)
    at Array.$__getAtomics (node_modules\mongoose\lib\types\array.js:260:35)
    at handleAtomics (node_modules\mongoose\lib\model.js:445:11)
    at model.Model.$__delta (node_modules\mongoose\lib\model.js:561:7)
    at model.Model.$__handleSave (node_modules\mongoose\lib\model.js:184:22)
    at model.Model.$__save (node_modules\mongoose\lib\model.js:230:8)
    at node_modules\kareem\index.js:276:8
    at _next (node_modules\kareem\index.js:102:16)
    at process.nextTick (node_modules\kareem\index.js:431:38)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

**If the current behavior is a bug, please provide the steps to reproduce.**
Run this script: https://gist.github.com/GilbertGobbels/9581be96cc94301d09ac2d0e4a45b290
It works whenever;
1. You grab a document (that is not new)
2. You push a new subdocument to an array of said document
3. You async save that document
4. You save that document again at just the right time that the atomics have been jsonified, but the document has not been reset yet

**What is the expected behavior?**
That some form of error handling within SubdocumentArray.toObject() that first checks if the toObject method exists, and if it does not, assumes it already IS an object, prevents the process from crashing AND the save from failing. Alternatively, check the constructor to confirm that it already IS a raw object.

**Please mention your node.js, mongoose and MongoDB version.**
8.10.0, 5.10.17, the MongoDB version does not impact this issue.

**Comment by lineus** [comment]
@GilbertGobbels see [this comment](https://github.com/Automattic/mongoose/issues/4064#issuecomment-380521818) on #4064. 

**TL;DR** Calling `.save()` on the same doc more than once in parallel might eventually throw a more specific error and generally speaking should be avoided.
